File: src/bigtube/core/scheduled_downloads.py
import time
from pathlib import Path
from typing import Any

from gi.repository import GLib

from .json_store import load_json, save_json
from .logger import get_logger

logger = get_logger(__name__)


class ScheduledDownloadStore:
    """Persists scheduled downloads so they survive application restarts."""

    _CONFIG_DIR = Path(GLib.get_user_config_dir()) / "bigtube"
    _FILE_PATH = _CONFIG_DIR / "scheduled_downloads.json"

    @classmethod
    def load(cls) -> list[dict[str, Any]]:
        data = load_json(cls._FILE_PATH, [])
        if not isinstance(data, list):
            return []
        return [item for item in data if isinstance(item, dict)]

    @classmethod
    def save(cls, items: list[dict[str, Any]]) -> None:
        if save_json(cls._FILE_PATH, items, indent=2):
            logger.debug("Scheduled downloads saved to disk")

    @classmethod
    def upsert(cls, item: dict[str, Any]) -> None:
        task_id = item.get("id")
        if not task_id:
            return

        items = [existing for existing in cls.load() if existing.get("id") != task_id]
        item = item.copy()
        item.setdefault("created_at", time.time())
        items.append(item)
        items.sort(key=lambda existing: existing.get("scheduled_time", 0))
        cls.save(items)

    @classmethod
    def remove(cls, task_id: str) -> None:
        if not task_id:
            return
        items = [item for item in cls.load() if item.get("id") != task_id]
        cls.save(items)

    @classmethod
    def clear_past(cls, now: float | None = None) -> list[dict[str, Any]]:
        now = time.time() if now is None else now
        due = []
        future = []
        for item in cls.load():
            if item.get("scheduled_time", 0) <= now:
                due.append(item)
            else:
                future.append(item)
        if due:
            cls.save(future)
        return due

File: rust/crates/bigtube-core/src/scheduled_downloads.rs
//! Persisted scheduled downloads, surviving restarts.
//! Ported from `core/scheduled_downloads.py`.

use std::path::PathBuf;

use serde_json::Value;

use crate::json_store::{load_json, save_json};
use crate::util::now_epoch;

pub struct ScheduledDownloadStore {
    path: PathBuf,
}

impl ScheduledDownloadStore {
    pub fn new(path: PathBuf) -> Self {
        Self { path }
    }

    /// Load, keeping only JSON objects (`load`).
    pub fn load(&self) -> Vec<Value> {
        let data: Value = load_json(&self.path, Value::Array(Vec::new()));
        match data {
            Value::Array(items) => items.into_iter().filter(Value::is_object).collect(),
            _ => Vec::new(),
        }
    }

    pub fn save(&self, items: &[Value]) {
        if save_json(&self.path, &items, Some(2)) {
            tracing::debug!("Scheduled downloads saved to disk");
        }
    }

    /// Insert or replace by `id`, default `created_at`, keep sorted by time.
    pub fn upsert(&self, item: &Value) {
        let task_id = match item.get("id").and_then(Value::as_str) {
            Some(id) if !id.is_empty() => id.to_string(),
            _ => return,
        };

        let mut items: Vec<Value> = self
            .load()
            .into_iter()
            .filter(|e| e.get("id").and_then(Value::as_str) != Some(task_id.as_str()))
            .collect();

        let mut item = item.clone();
        if let Value::Object(map) = &mut item {
            map.entry("created_at")
                .or_insert_with(|| Value::from(now_epoch()));
        }
        items.push(item);
        items.sort_by(|a, b| {
            let sa = a
                .get("scheduled_time")
                .and_then(Value::as_f64)
                .unwrap_or(0.0);
            let sb = b
                .get("scheduled_time")
                .and_then(Value::as_f64)
                .unwrap_or(0.0);
            sa.partial_cmp(&sb).unwrap_or(std::cmp::Ordering::Equal)
        });
        self.save(&items);
    }

    pub fn remove(&self, task_id: &str) {
        if task_id.is_empty() {
            return;
        }
        let items: Vec<Value> = self
            .load()
            .into_iter()
            .filter(|e| e.get("id").and_then(Value::as_str) != Some(task_id))
            .collect();
        self.save(&items);
    }

    /// Partition entries into due (<= now) and future; persist only the future
    /// ones and return the due ones (`clear_past`).
    pub fn clear_past(&self, now: Option<f64>) -> Vec<Value> {
        let now = now.unwrap_or_else(now_epoch);
        let mut due = Vec::new();
        let mut future = Vec::new();
        for item in self.load() {
            let t = item
                .get("scheduled_time")
                .and_then(Value::as_f64)
                .unwrap_or(0.0);
            if t <= now {
                due.push(item);
            } else {
                future.push(item);
            }
        }
        if !due.is_empty() {
            self.save(&future);
        }
        due
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn upsert_dedupes_and_sorts() {
        let dir = tempfile::tempdir().unwrap();
        let s = ScheduledDownloadStore::new(dir.path().join("sched.json"));
        s.upsert(&json!({"id": "a", "scheduled_time": 200.0}));
        s.upsert(&json!({"id": "b", "scheduled_time": 100.0}));
        s.upsert(&json!({"id": "a", "scheduled_time": 50.0})); // replace a
        let items = s.load();
        assert_eq!(items.len(), 2);
        // sorted by scheduled_time: a(50) then b(100)
        assert_eq!(items[0]["id"], json!("a"));
        assert_eq!(items[0]["scheduled_time"], json!(50.0));
        assert!(items[0].get("created_at").is_some());
    }

    #[test]
    fn clear_past_returns_due_and_keeps_future() {
        let dir = tempfile::tempdir().unwrap();
        let s = ScheduledDownloadStore::new(dir.path().join("sched.json"));
        s.upsert(&json!({"id": "old", "scheduled_time": 10.0}));
        s.upsert(&json!({"id": "new", "scheduled_time": 1000.0}));
        let due = s.clear_past(Some(100.0));
        assert_eq!(due.len(), 1);
        assert_eq!(due[0]["id"], json!("old"));
        let remaining = s.load();
        assert_eq!(remaining.len(), 1);
        assert_eq!(remaining[0]["id"], json!("new"));
    }
}
